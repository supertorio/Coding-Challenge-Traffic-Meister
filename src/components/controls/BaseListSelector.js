import React from 'react';
import PropTypes from 'prop-types';
import {withStyles} from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from "@material-ui/core/Typography";
import {List as ImmutableList} from 'immutable';

import {DEFAULT_SELECT_VALUE} from '../../constants/meister-consts';

const styles = theme => ({
    root: {
        width: '100%',
        position: 'relative',
    },
    listWrapper: {
        height: 200,
        overflow: 'auto',
        backgroundColor: theme.palette.background.paper,
    },
    optionText: {
        textTransform: 'capitalize'
    }
});

const BaseListSelector = (props) => {
    return (
        <div className={props.classes.root}>
            <Typography component="h2" variant="h5" gutterBottom>{props.label}</Typography>
            <div className={props.classes.listWrapper}>
                <List>
                    <ListItem
                        button
                        key={`item-${DEFAULT_SELECT_VALUE}`}
                        onClick={() => props.onChange(DEFAULT_SELECT_VALUE)}
                        selected={props.value === DEFAULT_SELECT_VALUE}>
                        <ListItemText primary={`Any`} />
                    </ListItem>
                    {props.options.map((option) => (
                        <ListItem
                            button
                            key={`item-${option}`}
                            onClick={() => props.onChange(option)}
                            selected={props.value === option}>
                            <ListItemText className={props.classes.optionText} primary={option}/>
                        </ListItem>
                    ))}
                </List>
            </div>
        </div>
    );
};

BaseListSelector.propTypes = {
    label: PropTypes.string.isRequired,
    classes: PropTypes.object.isRequired,
    value: PropTypes.string,
    options: PropTypes.instanceOf(ImmutableList),
    onChange: PropTypes.func,
};

BaseListSelector.defaultProps = {
    label: '',
    value: DEFAULT_SELECT_VALUE,
    options: new ImmutableList(),
    onChange: () => true,
};

export default withStyles(styles)(BaseListSelector);
