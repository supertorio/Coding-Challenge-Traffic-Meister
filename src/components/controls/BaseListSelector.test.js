import React from 'react';
import {BaseListSelector} from './BaseListSelector';
import {List} from 'immutable';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

describe('<BaseListSelector />', () => {

    const classesMock = {};
    let options;

    beforeEach(()=> {
        options = new List(['opt a', 'opt b', 'opt c']);
    });

    it('should match the snapshot', () => {
        const component = renderer.create(
            <BaseListSelector classes={classesMock} />,
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should display the provided options', () => {
        const wrapper = shallow(
            <BaseListSelector label="Type"
                              options={options}
                              classes={classesMock}
                              onChange={() => true} />
        );

        expect(wrapper.find(`ListItem[key="item-${options[0]}"]`).exists());
        expect(wrapper.find(`ListItem[key="item-${options[1]}"]`).exists());
        expect(wrapper.find(`ListItem[key="item-${options[2]}"]`).exists());
    });

    it('should react to selection change', () => {
        const selectionMock = jest.fn();
        const wrapper = mount(
            <BaseListSelector label="Type"
                              options={options}
                              classes={classesMock}
                              onChange={selectionMock} />
        );

        wrapper.find('ListItem').at(1).simulate('click');
        expect(selectionMock.mock.calls.length).toBe(1);
        expect(selectionMock.mock.calls[0][0]).toBe('opt a');
    });

});
