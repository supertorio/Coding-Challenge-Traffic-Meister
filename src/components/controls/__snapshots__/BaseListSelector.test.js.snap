// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseListSelector /> should match the snapshot 1`] = `
<div>
  <h2
    className="MuiTypography-root-1 MuiTypography-h5-17 MuiTypography-gutterBottom-28"
  >
    
  </h2>
  <div>
    <ul
      className="MuiList-root-37 MuiList-padding-38"
    >
      <div
        className="MuiButtonBase-root-53 MuiListItem-root-41 MuiListItem-default-44 MuiListItem-gutters-49 MuiListItem-button-50 MuiListItem-selected-52"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <div
          className="MuiListItemText-root-56"
        >
          <span
            className="MuiTypography-root-1 MuiTypography-subheading-8 MuiListItemText-primary-59"
          >
            Any
          </span>
        </div>
        <span
          className="MuiTouchRipple-root-62"
        />
      </div>
    </ul>
  </div>
</div>
`;
